src/lib/stores/theme.svelte.ts

```diff
@@ -0,0 +1,55 @@
+import { browser } from '$app/environment'
+import { changeColors, getEffectiveColorMode, type Mode } from '$lib/utils/colors'
+
+interface Theme {
+	baseColor: string
+	mode: Mode
+}
+
+function updateColors(baseColor: string, mode: Mode) {
+	const effectiveMode = getEffectiveColorMode(mode)
+	changeColors(baseColor, effectiveMode === 'dark')
+}
+
+export function withThemeStore(): Theme {
+	let baseColor = $state('#fefefe')
+	let mode = $state('system') as Mode
+
+	return {
+		get baseColor() {
+			return baseColor
+		},
+		set baseColor(value: string) {
+			baseColor = value
+			updateColors(baseColor, mode)
+		},
+		get mode() {
+			return mode
+		},
+		set mode(value: Mode) {
+			mode = value
+			updateColors(baseColor, mode)
+		},
+	}
+}
+
+function initializeTheme() {
+	const theme = withThemeStore()
+
+	// this would be better to do with CSS media queries to avoid initial flash
+	const effectiveMode = getEffectiveColorMode(theme.mode)
+	changeColors(theme.baseColor, effectiveMode === 'dark')
+
+	if (browser) {
+		const darkModePreference = window.matchMedia('(prefers-color-scheme: dark)')
+		// recommended method for newer browsers: specify event-type as first argument
+		darkModePreference.addEventListener('change', () => {
+			if (theme.mode !== 'system') {
+				return
+			}
+			updateColors(theme.baseColor, theme.mode)
+		})
+	}
+}
+
+initializeTheme()
```

Comment:
I am not quite sure why this creates a theme store instance and does not expose it. Maybe the intention was to have `theme` as a singleton and not export `withThemeStore`?

Another problem here is that the `change` event listener is never unsubscribed. Sorry I completely forgot I already worked on this im another project. While you can't use `$effect` in store, you can use `onDestroy`. https://github.com/vojtechsimetka/grans-game/blob/main/src/lib/stores/theme.svelte.ts

Reply:
The idea is that `initializeTheme` only runs once, detects the theme supported by the browser and sets the colors accordingly (line 39-41), then sets up an event listener if the browser theme changes. 

It creates a theme store instance so that it can use the values in the callback of the event listener.

There is no need to unsubscribe the event listener, because there is only one event listener ever created and it is needed for the whole lifetime when the page is open.

Reply:
Ok but the issue I see is that the `baseColor` and `mode` are not global variables. So within this `initializeTheme` function, there is a `theme` store, which has always default values (`baseColor` is white, `mode` is `system`). So if I change my the `baseColor` in the layout, then I change my browser prefferences from `light` to `dark`, the page still uses default  `baseColor`.

Reply:
I suggest two options:

1. Call `initializeTheme` in the `withThemeStore` and pass in the `theme` instance (`this`)
2. Change it to singleton, don't export the `withThemeStore`, instead `export const theme = withThemeStore()`

Reply:
Went with option 2.